From 84c29df01e985fb61cfcabe0e9a643981ce8c3f5 Mon Sep 17 00:00:00 2001
From: skyuv <skyuv@localhost>
Date: Wed, 12 Aug 2026 16:39:26 +0800
Subject: [PATCH] Fix C11 portability for variable stack buffers

Replace C99 variable-length arrays with alloca-backed buffers so MSVC can compile the Actor core without changing buffer lifetime.
---
 skynet-src/skynet_log.c    | 2 +-
 skynet-src/skynet_module.c | 4 ++--
 skynet-src/skynet_server.c | 8 ++++----
 3 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/skynet-src/skynet_log.c b/skynet-src/skynet_log.c
index 4ce106f..f62cf3a 100644
--- a/skynet-src/skynet_log.c
+++ b/skynet-src/skynet_log.c
@@ -11,7 +11,7 @@ skynet_log_open(struct skynet_context * ctx, uint32_t handle) {
 	if (logpath == NULL)
 		return NULL;
 	size_t sz = strlen(logpath);
-	char tmp[sz + 16];
+	char *tmp = alloca(sz + 16);
 	sprintf(tmp, "%s/%08x.log", logpath, handle);
 	FILE *f = fopen(tmp, "ab");
 	if (f) {
diff --git a/skynet-src/skynet_module.c b/skynet-src/skynet_module.c
index 9a8d618..8e18d92 100644
--- a/skynet-src/skynet_module.c
+++ b/skynet-src/skynet_module.c
@@ -32,7 +32,7 @@ _try_open(struct modules *m, const char * name) {
 	int sz = path_size + name_size;
 	//search path
 	void * dl = NULL;
-	char tmp[sz];
+	char *tmp = alloca(sz);
 	do
 	{
 		memset(tmp,0,sz);
@@ -78,7 +78,7 @@ static void *
 get_api(struct skynet_module *mod, const char *api_name) {
 	size_t name_size = strlen(mod->name);
 	size_t api_size = strlen(api_name);
-	char tmp[name_size + api_size + 1];
+	char *tmp = alloca(name_size + api_size + 1);
 	memcpy(tmp, mod->name, name_size);
 	memcpy(tmp+name_size, api_name, api_size+1);
 	char *ptr = strrchr(tmp, '.');
diff --git a/skynet-src/skynet_server.c b/skynet-src/skynet_server.c
index 782abff..a2fff2c 100644
--- a/skynet-src/skynet_server.c
+++ b/skynet-src/skynet_server.c
@@ -428,8 +428,8 @@ cmd_query(struct skynet_context * context, const char * param) {
 static const char *
 cmd_name(struct skynet_context * context, const char * param) {
 	int size = strlen(param);
-	char name[size+1];
-	char handle[size+1];
+	char *name = alloca(size+1);
+	char *handle = alloca(size+1);
 	sscanf(param,"%s %s",name,handle);
 	if (handle[0] != ':') {
 		return NULL;
@@ -478,7 +478,7 @@ cmd_kill(struct skynet_context * context, const char * param) {
 static const char *
 cmd_launch(struct skynet_context * context, const char * param) {
 	size_t sz = strlen(param);
-	char tmp[sz+1];
+	char *tmp = alloca(sz+1);
 	strcpy(tmp,param);
 	char * args = tmp;
 	char * mod = strsep(&args, " \t\r\n");
@@ -500,7 +500,7 @@ cmd_getenv(struct skynet_context * context, const char * param) {
 static const char *
 cmd_setenv(struct skynet_context * context, const char * param) {
 	size_t sz = strlen(param);
-	char key[sz+1];
+	char *key = alloca(sz+1);
 	int i;
 	for (i=0;param[i] != ' ' && param[i];i++) {
 		key[i] = param[i];
-- 
2.53.0.windows.3

